src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.Windows.cs

```diff
@@ -0,0 +1,336 @@
+                Interop.Crypt32.ContentType.CERT_QUERY_CONTENT_CERT;
+            const Interop.Crypt32.ExpectedContentTypeFlags ExpectedContentType =
+                Interop.Crypt32.ExpectedContentTypeFlags.CERT_QUERY_CONTENT_FLAG_CERT;
+
+            bool loaded = Interop.Crypt32.CryptQueryObject(
+                objectType,
+                pvObject,
+                ExpectedContentType,
+                Interop.Crypt32.ExpectedFormatTypeFlags.CERT_QUERY_FORMAT_FLAG_ALL,
+                dwFlags: 0,
+                pdwMsgAndCertEncodingType: IntPtr.Zero,
+                out Interop.Crypt32.ContentType actualType,
+                pdwFormatType: IntPtr.Zero,
+                phCertStore: IntPtr.Zero,
+                phMsg: IntPtr.Zero,
+                out SafeCertContextHandle singleContext);
+
+            if (!loaded)
+            {
+                singleContext.Dispose();
+                throw Marshal.GetHRForLastWin32Error().ToCryptographicException();
+            }
+
+            // Since contentType is an input filter, actualType should not be possible to disagree.
+            //
+            // Since contentType is only CERT, singleContext should either be valid, or the
+            // function should have returned false.
+            if (actualType != ContentType || singleContext.IsInvalid)
+            {
+                singleContext.Dispose();
+                throw new CryptographicException();
+            }
+
+            CertificatePal pal = new CertificatePal(singleContext, deleteKeyContainer: false);
+            return pal;
+        }
+
+        private static SafeCertStoreHandle ImportPfx(
+            ref BagState bagState,
+            ReadOnlySpan<char> password,
+            X509KeyStorageFlags keyStorageFlags)
+        {
+            ArraySegment<byte> reassembled = bagState.ToPfx(password);
+            SafeCertStoreHandle storeHandle = ImportPfx(reassembled, password, keyStorageFlags);
+            CryptoPool.Return(reassembled);
+
+            return storeHandle;
+        }
+
+        private static unsafe SafeCertStoreHandle ImportPfx(
+            ReadOnlySpan<byte> data,
+            ReadOnlySpan<char> password,
+            X509KeyStorageFlags keyStorageFlags)
+        {
+            const int MaxStackPasswordLength = 64;
+            void* alloc = default;
+            Span<char> szPassword = stackalloc char[MaxStackPasswordLength + 1];
+            ReadOnlySpan<char> effectivePassword = szPassword;
+            Interop.Crypt32.PfxCertStoreFlags flags = MapKeyStorageFlags(keyStorageFlags);
+
+            try
+            {
+                if (password.Contains('\0'))
+                {
+                    effectivePassword = password;
+                }
+                else
+                {
+                    if (password.Length >= MaxStackPasswordLength)
+                    {
+                        alloc = NativeMemory.Alloc((uint)password.Length + 1, sizeof(char));
```

Comment:
This is "fine" but I am wondering if this is an approach that you prefer now. Typically, we would just allocate on the heap, if the password is too big for the stack and put the whole thing in a `fixed`. Something like


```C#
Span<char> szPassword = stackalloc char[MaxStackPasswordLength + 1];

if (password.Length > MaxStackPasswordLength)
{
    szPassword = new char[password.Length + 1];
}

fixed (char* pPassword = szPassword)
{
    password.CopyTo(szPassword);
    szPassword[password.Length] = '\0';

    // Do the thing
}
```

What do we gain using `NativeMemory`, and, is this now "preferred" than using the managed heap for situations like this?

Reply:
My thoughts on when it is best to use `NativeMemory` is when we need a fixed buffer to survive outside the life of a method call (like the life of a class instance) but that doesn't seem to be the case here.